# AWID3 Dataset Analysis

* **Author:** Patrik Goldschmidt (igoldschmidt@fit.vut.cz)
* **Project:** Network Intrusion Datasets: A Survey, Limitations, and Recommendations
* **Date:** 2024

This notebook briefly presents the structure of a CSV AWID3 dataset format. However, PCAP format is available as well.

In [1]:
import pandas as pd
import numpy as np
import os

pd.set_option('display.max_columns', None)

In [2]:
DATAPATH = '/data/disk2/awid3/CSV/'

In [3]:
# List folder contents
for root, dirs, files in os.walk(DATAPATH):
    if not os.path.basename(root) == "":
        print(f'{os.path.basename(root)}/')
    for file in sorted(files):
        print('   ', file)

12.Evil_Twin/
    Evil_Twin_0.csv
    Evil_Twin_1.csv
    Evil_Twin_10.csv
    Evil_Twin_11.csv
    Evil_Twin_12.csv
    Evil_Twin_13.csv
    Evil_Twin_14.csv
    Evil_Twin_15.csv
    Evil_Twin_16.csv
    Evil_Twin_17.csv
    Evil_Twin_18.csv
    Evil_Twin_19.csv
    Evil_Twin_2.csv
    Evil_Twin_20.csv
    Evil_Twin_21.csv
    Evil_Twin_22.csv
    Evil_Twin_23.csv
    Evil_Twin_24.csv
    Evil_Twin_25.csv
    Evil_Twin_26.csv
    Evil_Twin_27.csv
    Evil_Twin_28.csv
    Evil_Twin_29.csv
    Evil_Twin_3.csv
    Evil_Twin_30.csv
    Evil_Twin_31.csv
    Evil_Twin_32.csv
    Evil_Twin_33.csv
    Evil_Twin_34.csv
    Evil_Twin_35.csv
    Evil_Twin_36.csv
    Evil_Twin_37.csv
    Evil_Twin_38.csv
    Evil_Twin_39.csv
    Evil_Twin_4.csv
    Evil_Twin_40.csv
    Evil_Twin_41.csv
    Evil_Twin_42.csv
    Evil_Twin_43.csv
    Evil_Twin_44.csv
    Evil_Twin_45.csv
    Evil_Twin_46.csv
    Evil_Twin_47.csv
    Evil_Twin_48.csv
    Evil_Twin_49.csv
    Evil_Twin_5.csv
    Evil_Twin_50.csv
    E

As apparent, there are several folders split by the attack type with a lot of files. We will count them...

In [4]:
dirs_total = 0
files_total = 0

for root, dirs, files in os.walk(DATAPATH):
    if not os.path.basename(root) == "":
        dirs_total += 1

    files_total += len(list(files))

dirs_total, files_total

(13, 745)

So there are 13 different attack types with 745 files in total. Count the total number of packets.

In [5]:
# Count the total number of packets
!( find $DATAPATH -name '*.csv' -print0 | xargs -0 cat ) | wc -l

36963644


In [6]:
# Load a sample file to see how to the CSV data look like
SAMPLE_DATA = os.path.join(DATAPATH, '7.SSH/SSH_24.csv')

data_sample = pd.read_csv(SAMPLE_DATA)

/tmp/ipykernel_55809/2719384289.py:4: DtypeWarning: Columns (71,73,77,78,79,80,82,84,85,87,90,91,92,93,98,99,100,101,102,103,104,105,107,108,109,110,111,115,118,119,120,135,137,140,141,142,144,146,148,149,151,166,167,168,178,180,183,184,185,190,192,193,194,195,196,197,200,201,202,203,204,205,206,209,210,211,212,213,215,217,218,219,247,248,249,250) have mixed types. Specify dtype option on import or set low_memory=False.
  data_sample = pd.read_csv(SAMPLE_DATA)


In [7]:
data_sample.info(verbose=True, show_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 254 columns):
 #    Column                                                          Non-Null Count  Dtype  
---   ------                                                          --------------  -----  
 0    frame.encap_type                                                50000 non-null  int64  
 1    frame.len                                                       50000 non-null  int64  
 2    frame.number                                                    50000 non-null  int64  
 3    frame.time                                                      50000 non-null  object 
 4    frame.time_delta                                                50000 non-null  float64
 5    frame.time_delta_displayed                                      50000 non-null  float64
 6    frame.time_epoch                                                50000 non-null  int64  
 7    frame.time_relative                   

In [8]:
len(data_sample)

50000

In [9]:
data_sample.head(10)

frame.encap_type  frame.len  frame.number  \
0                23         88       1200001   
1                23       1618       1200002   
2                23         88       1200003   
3                23         76       1200004   
4                23         70       1200005   
5                23        170       1200006   
6                23         88       1200007   
7                23         76       1200008   
8                23         76       1200009   
9                23         76       1200010   

                            frame.time  frame.time_delta  \
0  Dec 15- 2020 22:44:45.683738000 EET          0.000003   
1  Dec 15- 2020 22:44:45.684133000 EET          0.000395   
2  Dec 15- 2020 22:44:45.684136000 EET          0.000003   
3  Dec 15- 2020 22:44:45.684688000 EET          0.000552   
4  Dec 15- 2020 22:44:45.684717000 EET          0.000029   
5  Dec 15- 2020 22:44:45.684861000 EET          0.000144   
6  Dec 15- 2020 22:44:45.684864000 EET          0.000003   
7  Dec 15- 2020 22:44:45.684951000 EET          0.000087   
8  Dec 15- 2020 22:44:45.685139000 EET          0.000188   
9  Dec 15- 2020 22:44:45.685553000 EET          0.000414   

   frame.time_delta_displayed  frame.time_epoch  frame.time_relative  \
0                    0.000003        1608065086           393.831706   
1                    0.000395        1608065086           393.832101   
2                    0.000003        1608065086           393.832104   
3                    0.000552        1608065086           393.832656   
4                    0.000029        1608065086           393.832685   
5                    0.000144        1608065086           393.832829   
6                    0.000003        1608065086           393.832832   
7                    0.000087        1608065086           393.832919   
8                    0.000188        1608065086           393.833107   
9                    0.000414        1608065086           393.833521   

   radiotap.channel.flags.cck  radiotap.channel.flags.ofdm  \
0                           0                            1   
1                           0                            1   
2                           0                            1   
3                           0                            1   
4                           0                            1   
5                           0                            1   
6                           0                            1   
7                           0                            1   
8                           0                            1   
9                           0                            1   

   radiotap.channel.freq  radiotap.datarate  radiotap.dbm_antsignal  \
0                   5180               24.0                    -222   
1                   5180                NaN                    -224   
2                   5180               24.0                    -106   
3                   5180               24.0                    -106   
4                   5180               24.0                    -226   
5                   5180                NaN                    -103   
6                   5180               24.0                    -226   
7                   5180               24.0                    -105   
8                   5180               24.0                    -103   
9                   5180               24.0                    -103   

   radiotap.length  radiotap.mactime radiotap.present.tsft radiotap.rxflags  \
0               56      7.019137e+09                 1-0-0       0x00000000   
1               64               NaN                 0-0-0       0x00000000   
2               56      7.019138e+09                 1-0-0       0x00000000   
3               56      7.019138e+09                 1-0-0       0x00000000   
4               56      7.019138e+09                 1-0-0       0x00000000   
5               64               NaN                 0-0-0       0x00000000   
6 

In [10]:
data_sample['Label'].value_counts()

Label
Normal    50000
Name: count, dtype: int64

In [11]:
data_sample.describe()

frame.encap_type     frame.len  frame.number  frame.time_delta  \
count           50000.0  50000.000000  5.000000e+04      50000.000000   
mean               23.0    428.204240  1.225000e+06          0.000177   
std                 0.0    630.343484  1.443390e+04          0.000709   
min                23.0     70.000000  1.200001e+06          0.000000   
25%                23.0     76.000000  1.212501e+06          0.000028   
50%                23.0     79.000000  1.225000e+06          0.000068   
75%                23.0    194.000000  1.237500e+06          0.000191   
max                23.0   3156.000000  1.250000e+06          0.038847   

       frame.time_delta_displayed  frame.time_epoch  frame.time_relative  \
count                50000.000000      5.000000e+04         50000.000000   
mean                     0.000177      1.608065e+09           398.076655   
std                      0.000709      2.532041e+00             2.503457   
min                      0.000000      1.608065e+09           393.831706   
25%                      0.000028      1.608065e+09           395.914947   
50%                      0.000068      1.608065e+09           398.106473   
75%                      0.000191      1.608065e+09           399.986502   
max                      0.038847      1.608065e+09           402.685482   

       radiotap.channel.flags.cck  radiotap.channel.flags.ofdm  \
count                     50000.0                      50000.0   
mean                          0.0                          1.0   
std                           0.0                          0.0   
min                           0.0                          1.0   
25%                           0.0                          1.0   
50%                           0.0                          1.0   
75%                           0.0                          1.0   
max                           0.0                          1.0   

       radiotap.channel.freq  radiotap.datarate  radiotap.dbm_antsignal  \
count                50000.0       34188.000000            50000.000000   
mean                  5180.0          22.931555             -147.485320   
std                      0.0           3.662143               52.271686   
min                   5180.0           6.000000             -241.000000   
25%                   5180.0          24.000000             -207.000000   
50%                   5180.0          24.000000             -106.000000   
75%                   5180.0          24.000000             -103.000000   
max                   5180.0          24.000000             -103.000000   

       radiotap.length  radiotap.mactime  radiotap.timestamp.ts  \
count     50000.000000      3.418800e+04           5.000000e+04   
mean         58.520640      7.023315e+09           2.728415e+09   
std           3.716419      2.525216e+06           2.503453e+06   
min          56.000000      7.019137e+09           2.724170e+09   
25%          56.000000      7.021141e+09           2.726253e+09   
50%          56.000000      7.023302e+09           2.728445e+09   
75%          64.000000      7.025253e+09           2.730325e+09   
max          64.000000      7.027991e+09           2.733024e+09   

       radiotap.vendor_oui  wlan.duration  wlan.analysis.kck  \
count                  0.0   50000.000000                0.0   
mean                   NaN     170.210380                NaN   
std                    NaN     239.679879                NaN   
min                    NaN       0.000000                NaN   
25%                    NaN      48.000000                NaN   
50%                    NaN     108.000000                NaN   
75%                    NaN     236.000000                NaN   
max                    NaN   17440.000000                NaN   

       wlan.analysis.kek  wlan.fc.frag  wlan.fc.order  wlan.fc.moredata  \
count                0.0  50000.000000        50000.0      50000.000000   
mean                 NaN      0.000020            0.0         

Essentially that's it. Time and continuity analysis was performed via capinfos utility in a command line outside this notebook.